#  🛠️ From SQLite to BigQuery and Looker Studio

## 🔍 Overview:
Welcome to this hands-on data engineering assignment! Here, you’ll learn to work with real-world tools and platforms used by data professionals. 

This assignment is a continuation of a previous analysis where we used SQLite to store COVID-19 outcomes data. Now, we’ll take it a step further by:
1. **Migrating Data**: Moving the SQLite data into Google BigQuery, a robust cloud-based data warehouse.
2. **Setting Up Cloud Services**: Configuring Google Cloud Platform (GCP) with service accounts, authentication, and permissions.
3. **Visualizing Insights**: Using Looker Studio to create an interactive and insightful dashboard.

## 🎯 Learning Goals:
By the end of this assignment, you will:
1. Understand how to migrate data from a local SQLite database to BigQuery.
2. Learn to configure and use GCP, including APIs and service accounts.
3. Create interactive dashboards and visualizations in Looker Studio.
4. Gain confidence in working with cloud-based analytics tools.

---

## 📋 Prerequisites:
Before starting, ensure you have:
1. A Google account to access GCP and Looker Studio.
2. Python installed on your system (version 3.8 or higher).
3. Basic knowledge of:
   - Python scripting
   - SQL queries
   - Cloud platforms (optional but helpful)

---

## 🗂️ Assignment Workflow:
Here’s what we’ll cover step by step:
1. **Set up Environment** 
   - Update python to 3.8 or above.
   - Set up dependencies and install required packages.
2. **Set Up Google Cloud Platform (GCP)**:
   - Create a GCP project.
   - Enable required APIs (BigQuery, Cloud Storage).
   - Set up service accounts and authentication.
3. **Data Migration**:
   - Export SQLite data to  pandas dataframe.
   - Load the data into BigQuery.
4. **Visualization**:
   - Connect BigQuery to Looker Studio.
   - Design interactive dashboards.
5. **Wrap-Up**:
   - Tips,Sharing, troubleshooting.
---

## 🔗 References
- [Google Cloud Platform (GCP)](https://cloud.google.com/)
- [SQLite Documentation](https://www.sqlite.org/docs.html)
- [BigQuery Documentation](https://cloud.google.com/bigquery/docs)
- [Looker Studio Overview](https://lookerstudio.google.com/)
- [Python SQLite Guide](https://docs.python.org/3/library/sqlite3.html)
- [Getting Started with BigQuery](https://cloud.google.com/bigquery/docs/quickstarts)
- [GCP Console](https://console.cloud.google.com/)

---
## 🎉 Let’s Begin!
# Step 1: Setting up Python environment
Let's take the necessary steps to make sure that your Python environment is properly configured and all required dependencies are installed and ready.

In [8]:
# Step 1: Check Python Version and Environment
import sys

print(f"✅ Your Python version: {sys.version}")

# Ensure compatibility
if sys.version_info >= (3, 8):
    print("🎉 Your Python version is compatible!")
else:
    print("❌ Your Python version is lower than 3.8. Please upgrade Python to version 3.8 or higher.")
    print("Follow this guide: https://realpython.com/installing-python/")


✅ Your Python version: 3.14.7 (v3.14.7:823f0323ee6, Aug  5 2026, 07:07:01) [Clang 21.0.0 (clang-2100.1.1.101)]
🎉 Your Python version is compatible!


## Output:
1. If your Python version is **3.8 or higher**: You’re good to go! ✅
2. If your Python version is lower than 3.8:
   - You need to upgrade Python. Follow this guide: [Upgrade Python](https://realpython.com/installing-python/).

---

### Troubleshooting Python Issues:
1. **Using Anaconda**:
   - Check your environment's Python version:
     ```bash
     conda list python
     ```
   - Update Python in your environment:
     ```bash
     conda install python=3.8
     ```
2. **Using Standard Python**:
   - Download the latest version of Python from the [official site](https://www.python.org/downloads/).
   - Install it and ensure it’s added to your system PATH.


In [9]:
import subprocess
import sys

# List of required libraries
required_libraries = ["pandas", "google-cloud-bigquery", "google-cloud-storage", "google-auth", "google-cloud-bigquery"]

# Function to check and install missing libraries
def check_and_install_library(library):
    try:
        __import__(library.split("-")[0])  # Check if the library is installed
        print(f"✅ Library '{library}' is already installed.")
    except ImportError:
        print(f"❌ Library '{library}' is missing.")
        print(f"Installing '{library}'...")
        try:
            subprocess.check_call([sys.executable, "-m", "pip", "install", library])
            print(f"🎉 Successfully installed '{library}'!")
        except Exception as e:
            print(f"🚨 Error installing '{library}': {e}")
            print("👉 Try running the following command in your terminal:")
            print(f"pip install {library}")

# Check and install libraries
for lib in required_libraries:
    check_and_install_library(lib)


✅ Library 'pandas' is already installed.
✅ Library 'google-cloud-bigquery' is already installed.
✅ Library 'google-cloud-storage' is already installed.
✅ Library 'google-auth' is already installed.
✅ Library 'google-cloud-bigquery' is already installed.


## 🖥️ Understanding the Output:
1. ✅ If a library is installed, you’ll see a success message.
2. ❌ If a library is missing, it will be installed automatically.

---

### Troubleshooting Missing Libraries:
1. **Permission Issues**:
   - Add `--user` to the install command:
     ```bash
     pip install google-cloud-bigquery --user
     ```
2. **Environment Issues**:
   - Ensure you’re in the correct environment if using Anaconda:
     ```bash
     conda activate <your-environment-name>
     ```
3. **SSL Certificate Errors**:
   - Upgrade `pip` to fix SSL issues:
     ```bash
     pip install --upgrade pip
     ```
4. **Force Install**:
   - Reinstall the library with this command:
     ```bash
     pip install --force-reinstall google-cloud-bigquery
     ```

Still facing issues? Refer to these official library links for more help:
- [pandas Documentation](https://pandas.pydata.org/)
- [Google Cloud BigQuery Docs](https://cloud.google.com/bigquery/docs)
- [Google Cloud Storage Docs](https://cloud.google.com/storage/docs)


In [10]:
missing_libraries = []
for library in required_libraries:
    try:
        __import__(library.split("-")[0])
        print(f"✅ Library '{library}' is successfully installed.")
    except ImportError:
        missing_libraries.append(library)
        print(f"❌ Library '{library}' is still missing.")

if missing_libraries:
    print("\n🚨 The following libraries are still missing:")
    print(f"pip install {' '.join(missing_libraries)}")
else:
    print("\n🎉 All libraries are installed and verified!")


✅ Library 'pandas' is successfully installed.
✅ Library 'google-cloud-bigquery' is successfully installed.
✅ Library 'google-cloud-storage' is successfully installed.
✅ Library 'google-auth' is successfully installed.
✅ Library 'google-cloud-bigquery' is successfully installed.

🎉 All libraries are installed and verified!


# Step 2: Setting Up Google Cloud Platform (GCP)

## Why Set Up GCP?
To use BigQuery, Google Cloud Platform (GCP) is essential. It allows us to:
1. Organize resources using a **GCP Project**.
2. Enable the **BigQuery API** to interact with our data warehouse.
3. Use a **Service Account** for secure authentication when uploading data.

### 🔗 References:
- [Google Cloud Documentation](https://cloud.google.com/docs)
- [What is BigQuery?](https://cloud.google.com/bigquery/docs/introduction)
- [How to Use Service Accounts](https://cloud.google.com/iam/docs/understanding-service-accounts)
- [BigQuery Pricing](https://cloud.google.com/bigquery/pricing) (to understand usage costs)


## 🌐 Creating a GCP Project

A **GCP Project** is like a folder where you keep all your Google Cloud resources (e.g., APIs, datasets). It helps in organizing and managing access, billing, and usage.

📍 You’ll do all of this from the [BigQuery Console](https://console.cloud.google.com/bigquery).

### How to Create a GCP Project:
1. Go to the [BigQuery Console](https://console.cloud.google.com/bigquery).
2. Log in with your Google account.
3. In the top navigation bar, click **"Select a project"**.
4. Click **"New Project"**.
5. Fill out the form:
   - **Project Name**: Choose something meaningful, e.g., `SQLite-to-BigQuery-Project`.
   - **Organization**: If prompted, leave this as default.
6. Click **"Create"**.

### Output:
1. After creating the project, you’ll see it listed in your project dropdown(Select a project).
2. Note down your **Project ID** (you’ll use it later).
---

### 🔗 References:
- [How to Create a GCP Project](https://cloud.google.com/resource-manager/docs/creating-managing-projects)
- [GCP Project Overview](https://cloud.google.com/resource-manager/docs/overview)


## 📡Enabling BigQuery API

### Why Enable the API?
An **API** (Application Programming Interface) allows tools like Python to interact with GCP services. We need to enable the **BigQuery API** to:
1. Create tables in BigQuery.
2. Upload data from Python scripts.
---
### [Big Query](https://console.cloud.google.com/apis/library/bigquery.googleapis.com) should be enabled by default, if its not follow these steps

### How to Enable BigQuery API:
1. Open the [Google Cloud Console API Library](https://console.cloud.google.com/apis/library).
2. Ensure your project is selected in the top navigation bar.
3. Search for **BigQuery API** in the search bar.
4. Click on **BigQuery API** and then click **Enable**.
Note: It should be enabled by default if its not follow the below steps


### Output:
1. The BigQuery API will appear in your project’s **Enabled APIs and Services** list.
2. You can confirm this by visiting the [API Dashboard](https://console.cloud.google.com/apis/dashboard).

---

### 🔗 References:
- [BigQuery API Documentation](https://cloud.google.com/bigquery/docs/reference/rest)
- [How to Enable APIs in GCP](https://cloud.google.com/apis/docs/enable-disable-apis)

## 🚀 How to Create a Dataset in BigQuery

After enabling BigQuery for your Google Cloud project, follow these steps to create a dataset.



1. Navigate to the **[BigQuery Console](https://console.cloud.google.com/bigquery)**.  
   - This is where you can manage datasets and queries.

2. Select Your Project
   - In the left-hand Explorer panel, find your project (e.g., `your-project-id`).
   - If your project isn’t visible, click the dropdown at the top-left and select your project.

3. Click on the three dots next to your project id(`your-project-id`) and click on **"Create dataset"**

4. Fill in the Dataset Details
   - **Dataset ID**: Enter a name for your dataset (e.g., `nyc_covid_outcomes`). This name will be used to reference your dataset in code and queries.  
   - **Data location**: Choose the location (e.g., `US`) under the Multi-region. This defines where your data will be stored.  

5. Click "Create Dataset"

6. You're Done!
   - Your dataset will now appear under your project in the Explorer panel.  
  You can now upload data or create tables within it.


### 🔗 References:
- [BigQuery Documentation](https://cloud.google.com/bigquery/docs) – Official guide to BigQuery features.
- [Python BigQuery SDK](https://cloud.google.com/python/docs/reference/bigquery/latest) – For Python-based interactions with BigQuery.


   ## 🔑 Setting Up a Service Account

### What is a Service Account?
A **Service Account** is like a digital assistant that allows your Python script to interact with GCP securely. It’s safer than using personal credentials.

---

### How to Create a Service Account:
1. Go to the [IAM & Admin Console](https://console.cloud.google.com/iam-admin/serviceaccounts).
2. Click **"Create Service Account"**.
3. Fill out the details:
   - **Service Account Name**: Use something descriptive like `bigquery-access`.
   - Click **Create and Continue**.
4. Assign the following roles:
   - **BigQuery Admin**: Full access to create and manage BigQuery tables.
   - Search or Scroll down to the BigQuery section and find the BigQuery Admin role.
5. Click **Done**.

---

### Download the Authentication Key:
1. In the Service Account list, click on the newly created Service Account.
2. Navigate to the keys section from the navbar.
2. Click **"Add Key" > "Create New Key"**.
3. Select **JSON** format.
4. Save the key file securely, e.g., `bigquery-key.json`. This is your key file for authentication.

---

### 🔗 References:
- [How to Create a Service Account](https://cloud.google.com/iam/docs/creating-managing-service-accounts)
- [Best Practices for Service Accounts](https://cloud.google.com/iam/docs/best-practices-service-accounts)
- [How to Manage Service Account Keys](https://cloud.google.com/iam/docs/creating-managing-service-account-keys)


In [11]:
#Verify GCP Authentication

from google.oauth2 import service_account
from google.cloud import bigquery

# Replace the key path variable with the path to your downloaded JSON key file
key_path = r"/Users/alejandrorojo/Documents/Cloud_Keys/nydb-sqlite-to-bigquery-prjct-423e3c51e071.json"

try:
    # Authenticate using the Service Account
    credentials = service_account.Credentials.from_service_account_file(key_path)
    client = bigquery.Client(credentials=credentials, project=credentials.project_id)
    print(f"🎉 Successfully authenticated! Project ID: {credentials.project_id}")
except Exception as e:
    print(f"❌ Authentication failed: {e}")
    print("👉 Troubleshooting Tips:")
    print("- Ensure the key file path is correct.")
    print("- Verify the Service Account has the BigQuery Admin role.")
    print("- Confirm the BigQuery API is enabled for your project.")

🎉 Successfully authenticated! Project ID: nydb-sqlite-to-bigquery-prjct


## 🖥️ Understanding the Output:
1. **Success**:
   - If everything is set up correctly, you’ll see a confirmation with your Project ID.
2. **Failure**:
   - Authentication issues may occur if:
     - The JSON key file path is incorrect.
     - The Service Account lacks the correct roles.
     - The BigQuery API is not enabled.

---

## 🔧 Troubleshooting:
1. **Key File Issues**:
   - Ensure the JSON key file is in the correct location and hasn’t been renamed.
2. **Role Issues**:
   - Add the BigQuery Admin role to your Service Account using the [IAM Console](https://console.cloud.google.com/iam-admin/iam).
3. **API Issues**:
   - Revisit the [API Library](https://console.cloud.google.com/apis/library) and confirm the BigQuery API is enabled.

---

### 🔗 References:
- [Google Cloud Authentication](https://cloud.google.com/docs/authentication/production)
- [BigQuery Client Library Documentation](https://cloud.google.com/python/docs/reference/bigquery/latest)


# Step 3: Uploading Data from SQLite to BigQuery

## Why Upload Data to BigQuery?
BigQuery is a powerful data warehouse that allows us to:
1. Store large datasets in the cloud.
2. Query data using SQL with incredible speed and scalability.
3. Integrate with tools like Looker Studio for visualization.

In this step, we’ll:
1. Extract data from an SQLite database.
2. Transform it into a format BigQuery understands(Pandas DataFrame).
3. Upload it directly to a BigQuery table.

---

## 🔗  References
- [BigQuery Documentation](https://cloud.google.com/bigquery/docs)
- [Google BigQuery Python Client Library](https://cloud.google.com/python/docs/reference/bigquery/latest)
- [SQLite Documentation](https://www.sqlite.org/docs.html)

In [ ]:
#Import Required Libraries and Set Up Configuration
import pandas as pd
from sqlite3 import connect
from google.cloud import bigquery
from google.oauth2 import service_account
# ../nyc_covid_outcomes.db
# Set up the path to your SQLite database and BigQuery credentials
SQLITE_DB_FILE = r"/Users/alejandrorojo/Documents/DAEN_328/Assignments/Assignment_1/nyc_covid_outcomes.db"  # Replace with your SQLite database file from the first assignment
BIGQUERY_KEY_FILE = r"/Users/alejandrorojo/Documents/Cloud_Keys/nydb-sqlite-to-bigquery-prjct-423e3c51e071.json"  # Replace with your Service Account JSON key file that you just downloaded
PROJECT_ID = "nydb-sqlite-to-bigquery-prjct"  # Replace with your GCP Project ID
DATASET_ID = "nyc_covid_outcomes"  # Replace with your BigQuery dataset ID(this is the name of the dataset in the previous step)
# PROJECT_ID: 
# - Found in the GCP Console (https://console.cloud.google.com)
# - Navigate to the top-left corner, select the project (id must be displayed next to the project name or on the homescreen)
# DATASET_ID:
# - Found under "BigQuery" in the GCP Console.
# - Navigate to the BigQuery section (https://console.cloud.google.com/bigquery).
# - Select your project from the left panel, then select your dataset to view the dataset info 

###  Loading Data from SQLite:
1. Connect to the SQLite database.
2. Use a SQL query to extract all records from the `covid_outcomes` table.
3. Load the extracted data into a Pandas DataFrame for further processing.

In [2]:
# Load Data from SQLite into a Pandas DataFrame
def load_data_from_sqlite():
    """
    Load data from the 'covid_outcomes' table into a Pandas DataFrame.
    
    Returns:
        pd.DataFrame: Data loaded from the SQLite table.
    """
    try:
        # Connect to the SQLite database
        conn = connect(SQLITE_DB_FILE)
        query = "SELECT * FROM covid_outcomes"
        print("🔄 Loading data from the 'covid_outcomes' table...")
        data = pd.read_sql_query(query, conn)
        conn.close()
        print(f"✅ Successfully loaded {len(data)} records from 'covid_outcomes'.")
        return data
    except Exception as e:
        print(f"❌ Error loading data from 'covid_outcomes': {e}")
        return pd.DataFrame()


### Uploading Data to BigQuery
1. Authenticate with BigQuery using the Service Account key.
2. Define a schema for the BigQuery table (if it doesn’t already exist).
3. Upload the data from the DataFrame into the BigQuery table.


In [3]:
#Upload DataFrame to BigQuery
def upload_to_bigquery(dataframe, table_name):
    """
    Upload a Pandas DataFrame to a BigQuery table.
    
    Args:
        dataframe (pd.DataFrame): The DataFrame to upload.
        table_name (str): The name of the BigQuery table (e.g., "your_dataset_id.table_name").
    """
    try:
        # Authenticate and initialize the BigQuery client
        credentials = service_account.Credentials.from_service_account_file(BIGQUERY_KEY_FILE)
        client = bigquery.Client(credentials=credentials, project=PROJECT_ID)

        # Define table ID
        table_id = f"{PROJECT_ID}.{DATASET_ID}.{table_name}"
        
        # Configure the load job
        job_config = bigquery.LoadJobConfig(
            write_disposition=bigquery.WriteDisposition.WRITE_APPEND,  # Append to existing table
            create_disposition=bigquery.CreateDisposition.CREATE_IF_NEEDED  # Create table if not exists
        )

        # Start the upload job
        print(f"🔄 Uploading {len(dataframe)} records to BigQuery table: {table_id}")
        job = client.load_table_from_dataframe(dataframe, table_id, job_config=job_config)
        job.result()  # Wait for the job to complete
        print(f"✅ Successfully uploaded {len(dataframe)} records to {table_id}")
    except Exception as e:
        print(f"❌ Error uploading to BigQuery: {e}")


In [4]:
# Full Workflow
def push_to_bigquery():
    """
    Full workflow to load data from SQLite and upload to BigQuery.
    """
    print("🔄 Starting the workflow to upload data to BigQuery...")
    # Load data from SQLite
    dataframe = load_data_from_sqlite()
    
    if not dataframe.empty:
        # Specify the table name
        table_name = "covid_outcomes"  # Replace with your desired BigQuery table name
        
        # Upload data to BigQuery
        upload_to_bigquery(dataframe, table_name)
    else:
        print("⚠️ No data found in the 'covid_outcomes' table. Workflow skipped.")

# Execute the workflow
push_to_bigquery()

🔄 Starting the workflow to upload data to BigQuery...
🔄 Loading data from the 'covid_outcomes' table...
✅ Successfully loaded 176148 records from 'covid_outcomes'.
🔄 Uploading 176148 records to BigQuery table: nydb-sqlite-to-bigquery-prjct.nyc_covid_outcomes.covid_outcomes


/Users/alejandrorojo/Documents/DAEN_328/Assignments/daen328_env/lib/python3.14/site-packages/google/cloud/bigquery/_pandas_helpers.py:486: FutureWarning: Loading pandas DataFrame into BigQuery will require pandas-gbq package version 0.26.1 or greater in the future. Tried to import pandas-gbq and got: No module named 'pandas_gbq'
  warnings.warn(


✅ Successfully uploaded 176148 records to nydb-sqlite-to-bigquery-prjct.nyc_covid_outcomes.covid_outcomes


## 🖥️ Understanding the Output:
1. **Data Loaded Successfully**:
   - If the SQLite table contains data, you’ll see a message confirming the number of records loaded.
2. **Upload Status**:
   - A message confirms successful upload to BigQuery.
   - If any errors occur, detailed messages will help debug the issue.

---

## 🕵️ Exploring the Uploaded Data in BigQuery:
1. **Verify the Table in BigQuery**:
   - Open the **[BigQuery Console](https://console.cloud.google.com/bigquery)**.
   - Navigate to your project in the left-hand Explorer panel.
   - Expand the dataset where your table is uploaded (e.g., `nyc_covid_outcomes`).
   - Click on the table name (e.g., `covid_outcomes`) to open it.

2. **Preview the Data**:
   - In the table view, click the **"Preview"** tab to view the first few rows of your data. This is a quick way to check if the upload was successful.
   - Check the **"Schema"** tab to verify the column names and data types.
3. **Run Queries**:
   - Click on the **"+"** tab and use SQL to explore your data. Example query:
     ```sql
     SELECT * 
     FROM `your-project-id.nyc_covid_outcomes.covid_outcomes`
     LIMIT 10;
     ```
   - Modify the query to filter or summarize the data as needed.



---


## 🔧 Troubleshooting:
1. **SQLite Issues**:
   - Ensure the database file exists in the specified path.
   - Verify the `covid_outcomes` table exists using a database viewer like DB Browser.
2. **BigQuery Issues**:
   - Ensure the Service Account JSON key is correctly configured.
   - Confirm the dataset exists in BigQuery.
   - Verify the Service Account has the necessary roles (BigQuery Admin).

---

### 🔗 References:
- [BigQuery Python Client](https://cloud.google.com/python/docs/reference/bigquery/latest)
- [Pandas Documentation](https://pandas.pydata.org/docs/)
- [SQLite Documentation](https://www.sqlite.org/docs.html)


# Step 4: Visualizing Data in Looker Studio

## Why Use Looker Studio?
[Looker Studio](https://lookerstudio.google.com/) allows us to:
1. Create interactive dashboards to visualize BigQuery data.
2. Share insights in a user-friendly format.
3. Enable real-time exploration of datasets.

In this step, we will:
1. Connect Looker Studio to BigQuery.
2. Build an interactive dashboard for the `covid_outcomes` table.

---

### 🔗 References:
- [Looker Studio Overview](https://lookerstudio.google.com/)
- [Connecting Looker Studio to BigQuery](https://support.google.com/looker-studio/answer/6370296?hl=en)
- [BigQuery Visualization Best Practices](https://cloud.google.com/bigquery/docs/visualize-looker)


## 📈 Connecting Looker Studio to BigQuery

Connecting Looker Studio to BigQuery allows us to directly query our dataset and create visualizations without exporting the data manually.

### How to Connect:
1. Open [Looker Studio](https://lookerstudio.google.com/).
2. Click **"Blank Report"** or **"Create"**.
3. In the **Add Data to Report** dialog:
   - Select **BigQuery**.
   - Click **Authorize** if prompted to grant access.
4. Choose your project:
   - Select your GCP project (e.g., `SQLite-to-BigQuery-Project`).
   - Navigate to the dataset (e.g., `nyc_covid_outcomes`).
   - Select the table (e.g., `covid_outcomes`).
5. Click **Add** and confirm by clicking **Add to Report**.

---

### Output:
Your BigQuery table will now appear as a data source in Looker Studio. You can start building visualizations using this data source.


## 📊 Creating Visualizations

### Why Create Visualizations?
Visualizations help us understand data better by showing patterns, trends, and comparisons. In this step, we’ll create easy-to-build charts to explore the `covid_outcomes` data in Looker Studio.

---

### Suggested Charts for `covid_outcomes` Table

#### 1. **Confirmed Cases Over Time**
- **Chart Type**: Line Chart
- **Purpose**: Show how the number of confirmed cases changed over time.
- **Steps**:
  1. Add a **Line Chart** from the toolbar.
  2. Set the **Dimension** to `specimen_date`.
  3. Set the **Metric** to `number_confirmed`.
  4. Rename the chart title to "Confirmed Cases Over Time" under the style tab.

---

#### 2. **Deaths Over Time**
- **Chart Type**: Line Chart
- **Purpose**: Show how the number of deaths changed over time.
- **Steps**:
  1. Duplicate the "Daily Confirmed Cases" chart (right-click > Duplicate).
  2. Replace the **Metric** with `number_deaths`.
  3. Rename the chart title to "Deaths Over Time."

---
#### 3. **Hospitalizations Over Time**
- **Chart Type**: Line Chart
- **Purpose**: Show how the number of Hospitalizations changed over time.
- **Steps**:
  1. Duplicate the "Daily Confirmed Cases" chart (right-click > Duplicate).
  2. Replace the **Metric** with `number_hospitalized`.
  3. Rename the chart title to "Hospitalizations Over Time."

---


#### 4. **Key Metrics Summary**
- **Chart Type**: Scorecards
- **Purpose**: Highlight total numbers for key metrics (e.g., total confirmed cases, total deaths).
- **Steps**:
  1. Add **4 Scorecards** from the toolbar.
  2. For each scorecard, set the **Metric** to one of the following:
     - `number_tested` 
     - `number_confirmed` 
     - `number_hospitalized` 
     - `number_deaths` 
  3. Rename the scorecards to "Total Tested," "Total Confirmed," etc.

---
### 5. Testing vs Confirmed Cases
- **Chart Type**: Combo Chart (Line and Bar)
- **Purpose**: Compare the number of tests conducted with confirmed cases over time.

#### Steps:
1. Add a **Combo Chart**.
2. Set the **Dimension** to `specimen_date`.
3. Add the following **Metrics**:
   - `number_tested` (Bar)
   - `number_confirmed` (Line)
4. Rename the chart title to **"Testing vs Confirmed Cases"**.

---
### 6. **Custom Calculations**
- **Purpose**: Create additional metrics to derive more insights from the existing data using calculated fields.
- **Usage**: 
  - In the **Data tab** (on the right-hand side), you can view the current schema of the data being used.
  - To create a custom value, click **Add a Field** → **Add a Calculated Field**.

#### **Examples of Custom Calculations**:
1. **Death Rate**:
   - Formula: `number_deaths / number_confirmed`
2. **Hospitalization Rate**:
   - Formula: `number_hospitalized / number_confirmed`
3. **Positive Test Rate**:
   - Formula: `number_confirmed / number_tested`

#### **Visualization**:
- After creating calculated fields, use them in visualizations like **Line Charts** or **Bar Charts**:
  1. Add a **Line Chart** or **Bar Chart**.
  2. Set the **Dimension** to `specimen_date`.
  3. Use the calculated values (e.g., Death Rate, Hospitalization Rate) as **Metrics**.
  4. Customize the chart title for clarity (e.g., "Death Rate Over Time").

### Output:
By the end of this step, your dashboard should have charts that provide a detailed overview of the COVID-19 data.


## 📤 Finalizing and Sharing

### How to Finalize:
1. Review all charts for accuracy and aesthetics.
2. Customize the layout and design:
   - Add a title and description for your dashboard.
   - Use consistent colors for metrics.
3. Create separate pages for sections of data.
4. Click **View** in the top-right corner to review the report's functionality and visual design.

---

### How to Share:
1. Click **Share** in the top-right corner.
2. Choose one of the following options:
   - **Invite People**: Add specific viewers by email.
   - **Get Link**: Create a shareable link (adjust permissions if needed).

---

## 🎉 Your dashboard is now ready to share and explore!


## 🔧 Troubleshooting Looker Studio

### Common Issues:
1. **BigQuery Data Not Loading**:
   - Ensure your Service Account has access to the dataset.
   - Verify the dataset and table names are correct.
2. **Chart Not Displaying Data**:
   - Check that dimensions and metrics are correctly assigned.
   - Ensure the data type matches the chart type (e.g., numeric for metrics).

---

### 🔗 References:
- [Looker Studio Support](https://support.google.com/looker-studio/)
- [Fixing Common Looker Studio Issues](https://support.google.com/looker-studio/answer/7020039)


# 📝 Step 5: Final Deliverables

You have migrated your data and built your dashboard. This last step is where you show us what it means. **This section is worth 10 points.**

---

## 🔗 Deliverable 3: Your Looker Studio Report Link

Paste your report's share link below. Before you do, two checks:

1. **Sharing**: In Looker Studio, click **Share** → set link access to **"Anyone with the link"** → **Viewer**.
2. **Data credentials**: Open your BigQuery data source → **Data credentials** → set to **Owner's credentials**. If this is left on *Viewer's credentials*, Looker Studio will try to query BigQuery using **our** account when we open your link — and we have no access to your project. Your report will render as an error and score as a report you did not build.

Verify by opening the link in a private/incognito window while signed out of your Google account.

**My Looker Studio report link:** `https://datastudio.google.com/reporting/34a14452-894d-426c-a8f4-cb99726a7f18`

---

## ✍️ Deliverable 4: Written Interpretation

### 👉 The questions are in **`Assignment02_Addendum.pdf`**, attached with this assignment.

That file contains **20 questions** in four sections:

| Section | Topic | Questions |
|---|---|---|
| 1 | GCP & BigQuery Basics | 1–5 |
| 2 | Setting Up BigQuery & Data Upload | 6–11 |
| 3 | Querying Data in BigQuery | 12–16 |
| 4 | Exploring, Validating, and Understanding Data | 17–20 |

### How to submit your answers

1. Open `Assignment02_Addendum.pdf`.
2. Add a **new markdown cell below this one**.
3. Answer **all 20 questions**, numbered to match the addendum so we can follow along.
4. For questions asking for SQL (12–14), format your query in a code block so it stays readable:

   ````
   ```sql
   SELECT COUNT(*) FROM `your-project-id.nyc_covid_outcomes.covid_outcomes`;
   ```
   ````

### What earns full marks

Answer from **your own project, your own numbers, and your own dashboard**. Many of these questions ask what *you* did and what *you* saw — answers that could have been written without doing the assignment will score in the "Generic description not tied to the student's own numbers" band of the rubric.

Two questions carry extra weight, because the rubric requires a data quality observation carried forward from Assignment 1:

- **Q18** — patterns and trends: describe what your charts actually show, with approximate dates, not what you expected to see.
- **Q19** — validation: this is where that Assignment 1 data quality issue belongs. Say whether it still appears in BigQuery and how it affects the way your dashboard should be read.

---

## ✅ Before you submit

- [ ] Row count in BigQuery reconciles with Assignment 1 (re-running the upload cell is the usual cause of a mismatch)
- [ ] All six visuals present, every time axis using `specimen_date` sorted ascending
- [ ] All three calculated fields defined and used in at least one chart
- [ ] All 20 addendum questions answered and numbered
- [ ] Share link opens correctly in a private browser window
- [ ] Notebook run top-to-bottom with outputs visible
- [ ] **No key file and no `.db` file in your GitHub repo** — run `git status` and confirm


**Addendum Answers**

Section 1:

1. SaaS, or Software as a Service provides complete applications that users access through the internet without having to manage underlying infrasttructure. For example, in this assignment we used Looker Studio (Data Studio) to create dashboards for our data. PaaS, or Platform as a Service, provides a managed enviroment for developing and deploying applications without managing servers. For example, the Google App Engine helps provide this service. Lastly, IaaS, or Infrastructure as a Service, provides virtualized computing resources such as virtual machines, storage, and networking. An example of this service would be Google Computer Engine.


2. BigQuery and traditional on-premises databases differ in many ways. BigQuery is a fully mangable, serverless cloud data warehouse, while traditional on-premises databases require organizations to maintain their own servers, storage, database software, backups, and capacity. BigQuery can automatically scale to handle large datasets and complex SQL queries without requiring us to manage underlying hardware. Some advantages that BigQuery has it, cloud accessiblity, integration with tools such as Looker studio, and the ability to pay for storage and processes as needed. These advantages can be seen in this assignment itself as we use BigQuery to host our local sqlite database created in assignment one, as well as connecitng cloud table to Looker Studio to create a dashboard of data discoveries and visualizations.

3. Enabling the BigQuery API before using it allows applications outside the Google Cloud Console, such as our code in this assignment, to communicate with BigQuery. It is critical to enable to API so that our python code could create and access BigQuery resoucres and in order to upload our database from assignment 1. 

4. A service account is a special Google Cloud identity intended for the application or program rather than a human user. This assignment requires it, as our python code could securly authenticate with GCP and receive permissions to access BigQuery. A regular service account represents a person and is normally used for interactive login, while in contrast, the a service account allows for python programs to authenticate and perform operations automatically.

5. I authenticated my Python script with BigQuery using the generated JSON key for my service account. I used the  `service_account.Credentials.from_service_account_file()` to load my credentials and then created a `bigquery.Client` using those credentials and associated project ID.

Section 2:

6. The steps I followed were as follows: Creating a GCP project and enabling the Big Query API, Opening the BigQuery console and selecting my created project and creating the dataset `nyc_covid_outcomes`. I then choose the dataset location through the explorer panel, used my python configuration by using what the names I chose for my project and dataset which are `nydb-sqlite-to-bigquery-prjct` and `nyc_covid_outcomes`. Then the `covid_outcomes` table was created during the upload due to oure configuration in our python code. Lastly, I then loaded the data into `nydb-sqlite-to-bigquery-prjct.nyc_covid_outcomes.covid_outcomes`.

7. The structure of the database is as follows: The database contains one table named `covid_outcomes` containing 176148 rowws and 6 coluumns as verifyed from assignment 1. The columns are `extract_date`, `speciman_date`, `number_tested`, `number_confirmed`, `number_hospitalized`, and `number_deaths`. In the SQLite database, `extract_date` and `speciman_date` were stored as strings, while the others were stored as integers. In BigQuery, the equivalent inferred types were strings for the two date fields and integers for the numerical fields.

8. The python libraries used were `google.cloud.bigquery` and `google.oauth2.service_account` to connect to BigQuery. `google.oauth2.service_account` was also used to load and authenticate the service account credentials from the JSON key file. `google.cloud.bigquery` was then also used to create the BigQuery client, connect it to my GCP project, and to upload data to the BigQuery table.

9. The steps invovled were as follows: I connected the SQLite database in the function `load_data_from_sqlite()`, in it the `connect()` ran and then the `SELECT * FROM covid_outcomes` ran. After the query results were loaded into a pandas dataframe using `pd.read_sql_query()` and then closing the SQLite connection. Next, I authenticated to BigQuery  with my service account credentials, created BigQuery client, and identified the destination table by using my project ID, dataset ID, and table name. Lastly, I used `client.load_table_from_dataframe()` to upload the dataframe to BigQuery and waited for the load job to finish with `job.result()`, which resulted in a message indicating that the upload was successful.

10. The challenges I faced when uploading the data were when I discoverd that my virtual enviroment is set up a bit funky. At the beginning before starting, I did not realize I installed pyarrow correctly, this resulting in the functions when uploading data to BigQuery to error out and giving a message I did not have pyarrow installed. After doing some reasearch I found out that there is two virtual enviroments that I have, except one is not showing and is pathed somewhere that I have no clue where it can be. Alongside that google recommended me to the command `python -m pip install pyarrow` as this command allowed for me to install pyarrow to the correct place. Overall the fix was relatively simple bur requires more digging on my side on why this occured. 

11. For a larger dataet, I would avoid loading the entirity of the SQLite table into one single dataframe at once. Instead I would process and upload the data in smaller chunmks or batches to reduce memory usage (I would probably refer back to batch uploading from Assignment 1 readings).  I would also use an incremental loading approach so that only new or changed records are uploaded instead of fully reuploading the entire dataset. Lastly, I would add checks to prevent duplicate data records in case of needing to rerun the pipeline in case errors or interuptions occur. 

Section 3:

12. 
`SELECT COUNT(*) AS total_rows
FROM 'nydb-sqlite-to-bigquery-prjct.nyc_covid_outcomes.covid_outcomes';`


13. 
`SELECT * FROM 'nydb-sqlite-to-bigquery-prjct.nyc_covid_outcomes.covid_outcomes'
LIMIT 10;`

14. 
I would do the following:
`SELECT *
FROM 'nydb-sqlite-to-bigquery-prjct.nyc_covid_outcomes.covid_outcomes'
WHERE DATE(specimen_date) BETWEEN DATE 'XXXX-XX-XX' AND DATE 'XXXX-XX-XX';`

15. The purpose of using LIMIT in a query is to restrict the number of rows returned by our query. I would use this when needing to preview or inspect a small smaple of the data instead of displaying the entire dataset. An example of this is when I used the same sql query as in question 13 to double check if the data got uploaded correctly without loading the entire table. 

16. To optimize an SQL query means writing it so it processes only the data needed to answer the question. In BigQuery, queries can take longer and cost more when they scan larger amounts of data. For example, using `SELECT *` reads every column, even if only one or two columns are needed. Applying filters and selcting only neccessary columns can reduce the amount of data processed.

Section 4:

17. I used Looker by directly connecting it to my BigQuery `covid_outcomes` table and using it as the data source for my dashboard. I then selected my GCP project, dataset, and BigQuery table and added it to the report. I then created the charts showing information on confirmed cases, hospitlizations over time, and much more.

18. The data analyzed seemed reasonable, as my dashboard showed the large first wave in spring 2020 and then slowly decreasing after the major spike. Confirmed cases peaked around April 2020 at about 124000 cases for the month, hospitalizations peaked at 31000. Another observation that was seen in the data was the noticeable increase in confirmed cases during late 202 and early 2021, reaching 92000 in January 2021. Although, hospitalizations during this second increase were much lower compared to spring 2020. In the calculated rates, is showed similar patterns, where the average postive test rate was high around March-April 202 at about 47%. It then decreased substantially during the summer and increased again around winter of 2020-2021, which then fell to reltively low level by summer 2021. Death and Hospitlization rates also declined after the first several months of the pandemic. Lastly, when creating graphs for the custom calculations I took the time to create an extra graph that compares the Death Rate and Hospitalization rate during the pandemic as I found it to be interesting to compare due to how news articles highligted how people being hospitlized by Covid led to hospitals being maxed out and making it hard for infected individuals to get care. When looking at this graph we can see that at Covids peak the death rate was at or above 10% with hospitalization rates above 20%, thus showing that the data did in fact back up the claims that due to hospitals being full, death rates were high as the people that died could have been people who were not able to get care in time to be saved.

19. I validated the data by comparing the number of records in SQLite database table with the number of records uploaded to BigQuery. Both of them contained 176148 rows of data, thus confirming that the full data transferred successfully. I also conmpared the fields and interpretations using the NYC Open Data Documentation that was downloaded in Assignment 1. The documention helped in ensuring I filtered or grouped data correctly as the documentation directly states to filter or group by extract_date for the most accurate up to date information as not doing so can cause values to be counted twice resulting in a data analysis full of errors. Furthermore, it is important to note that the data quality issues from assignment 1 also exists in BigQuery as the BigQuery table was made by the upload of the SQLite table. Thus, leading to some of the incorrect future dates of 2200 and above to still be in this dataset. This then led me to creating filters on my data for `speciman_date` and `extract_date` to only have valid date values between 2020-2021. 

20.  Yes, we could have skipped loading the data into an SQLite database locally. We could have retrieved an API in Assignment 1 and loaded it directly into BigQuery. The main advantage this skip provides is the amount of steps in the pipeline, e.g. our pipleline would be
API -> BigQuery instead of API -> SQLite -> Pandas -> BigQuery. This would also result in a reduction in the nubmer of data transfer steps, avoiding the maintainment of a local database file and making the process easier to automate and scale for larger datasets. However, the use and creation of an SQLite database has its own advantages asw well. The SQLite database provides us with not only a local copy of the raw data but it also allows use to inspect, query, and validate the data before uploading it to the cloud. It also serves as a checkpoint between the API and BigQuery, so in case something goes wrong during the BigQuery upload, we do not have to retrieve the API data again. To add-on although there are different pros to each method, the main disadvantage of uploading directly to BigQuery is that we would lose an intermediate staging and validation step and would need to add more data validation and error handling steps to ensure data is not malformed or uploaded incorrectly. Although uploading the data would be more efficient, the staging and validation steps lost are just to valuable to lose, thus even if directly uploading the data is more efficient, it is better practice to have a local copy of the data if possible to just have a cushion of comfort in case things end up going wrong. 


